File: src/selectors.test.ts
import {
	clampIndex,
	firstSectionOfTab,
	listStatus,
	promptSegments,
	visibleSections
} from './selectors'
import { SECTIONS } from './content'

test('with no filter, shows only the active tab', () => {
	expect(visibleSections('projects', '', false).map((s) => s.id)).toEqual([
		'mapwright',
		'3dpass',
		'transfer',
		'platform',
		'hat',
		'oss'
	])
	expect(visibleSections('work', '', false).map((s) => s.id)).toEqual(['mapwright-work', 'c1', 'stack'])
})

test('while the filter is open, searches every section regardless of tab', () => {
	const hits = visibleSections('readme', 'stack', true).map((s) => s.id)
	expect(hits).toContain('stack')
})

test('with a filter string but the filter closed, stays within the active tab', () => {
	expect(visibleSections('readme', 'stack', false).map((s) => s.id)).toEqual([])
})

test('filter matches label and title, case-insensitively', () => {
	expect(visibleSections('projects', 'MAPWRIGHT', true).map((s) => s.id)).toEqual([
		'mapwright',
		'mapwright-work'
	])
	expect(visibleSections('projects', 'secure element', true).map((s) => s.id)).toEqual(['hat'])
})

test('filter with no match returns nothing', () => {
	expect(visibleSections('projects', 'zzzzqqq', true)).toEqual([])
})

test('list status reports position and total', () => {
	const visible = visibleSections('projects', '', false)
	expect(listStatus(visible, 'mapwright', '')).toBe('1/6')
	expect(listStatus(visible, 'oss', '')).toBe('6/6')
})

test('list status marks an active filter', () => {
	const visible = visibleSections('projects', 'map', true)
	expect(listStatus(visible, 'mapwright', 'map')).toBe('1/2 filtered')
})

test('list status reports no match on an empty list', () => {
	expect(listStatus([], 'mapwright', 'zzz')).toBe('no match')
})

test('prompt always carries directory and branch', () => {
	const readme = SECTIONS.find((s) => s.id === 'readme')!
	const segs = promptSegments(readme)
	expect(segs[0].text).toBe('~/site/README.md')
	expect(segs[1].text).toContain('main')
	expect(segs).toHaveLength(2)
})

test('prompt adds a language module only for sections that define one', () => {
	const mapwright = SECTIONS.find((s) => s.id === 'mapwright')!
	expect(promptSegments(mapwright)).toHaveLength(3)
	expect(promptSegments(mapwright)[2].text).toContain('docker')

	const contact = SECTIONS.find((s) => s.id === 'contact')!
	expect(promptSegments(contact)).toHaveLength(2)
})

test('first section of a tab', () => {
	expect(firstSectionOfTab('projects').id).toBe('mapwright')
	expect(firstSectionOfTab('work').id).toBe('mapwright-work')
})

test('clampIndex stops at both ends and does not wrap', () => {
	expect(clampIndex(0, -1, 5)).toBe(0)
	expect(clampIndex(4, 1, 5)).toBe(4)
	expect(clampIndex(0, 99, 5)).toBe(4)
	expect(clampIndex(4, -99, 5)).toBe(0)
	expect(clampIndex(1, 1, 5)).toBe(2)
})

File: internal/tui/selectors_test.go
package tui

import (
	"testing"

	site "github.com/phetzy/fetzycloudonline"
)

func ids(sections []site.Section) []string {
	out := make([]string, 0, len(sections))
	for _, s := range sections {
		out = append(out, s.ID)
	}
	return out
}

func equal(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func TestVisibleSectionsNoFilterShowsActiveTab(t *testing.T) {
	c := site.MustLoad()
	got := ids(VisibleSections(c, "projects", "", false))
	want := []string{"mapwright", "3dpass", "transfer", "platform", "hat", "oss"}
	if !equal(got, want) {
		t.Errorf("got %v, want %v", got, want)
	}
}

func TestVisibleSectionsFilterOpenSearchesEveryTab(t *testing.T) {
	c := site.MustLoad()
	got := ids(VisibleSections(c, "readme", "stack", true))
	if !equal(got, []string{"stack"}) {
		t.Errorf("got %v, want [stack] — the filter must cross tabs while open", got)
	}
}

func TestVisibleSectionsFilterClosedStaysInTab(t *testing.T) {
	c := site.MustLoad()
	got := ids(VisibleSections(c, "readme", "stack", false))
	if len(got) != 0 {
		t.Errorf("got %v, want none — a closed filter stays within the active tab", got)
	}
}

func TestVisibleSectionsMatchesCaseInsensitively(t *testing.T) {
	c := site.MustLoad()
	if got := ids(VisibleSections(c, "projects", "MAPWRIGHT", true)); !equal(got, []string{"mapwright"}) {
		t.Errorf("got %v, want [mapwright]", got)
	}
	if got := ids(VisibleSections(c, "projects", "secure element", true)); !equal(got, []string{"hat"}) {
		t.Errorf("got %v, want [hat] — matching covers title as well as label", got)
	}
}

func TestListStatus(t *testing.T) {
	c := site.MustLoad()
	visible := VisibleSections(c, "projects", "", false)

	if got := ListStatus(visible, "mapwright", ""); got != "1/6" {
		t.Errorf("got %q, want %q", got, "1/6")
	}
	if got := ListStatus(visible, "oss", ""); got != "6/6" {
		t.Errorf("got %q, want %q", got, "6/6")
	}
	filtered := VisibleSections(c, "projects", "map", true)
	if got := ListStatus(filtered, "mapwright", "map"); got != "1/1 filtered" {
		t.Errorf("got %q, want %q", got, "1/1 filtered")
	}
	if got := ListStatus(nil, "mapwright", "zzz"); got != "no match" {
		t.Errorf("got %q, want %q", got, "no match")
	}
}

func TestPromptSegments(t *testing.T) {
	c := site.MustLoad()
	var readme, mapwright, contact site.Section
	for _, s := range c.Sections {
		switch s.ID {
		case "readme":
			readme = s
		case "mapwright":
			mapwright = s
		case "contact":
			contact = s
		}
	}

	if got := PromptSegments(readme); len(got) != 2 {
		t.Errorf("readme: got %d segments, want 2 (directory and branch)", len(got))
	}
	if got := PromptSegments(contact); len(got) != 2 {
		t.Errorf("contact: got %d segments, want 2 — no language module", len(got))
	}
	got := PromptSegments(mapwright)
	if len(got) != 3 {
		t.Fatalf("mapwright: got %d segments, want 3", len(got))
	}
	if got[0].Text != "~/site/projects/mapwright" {
		t.Errorf("first segment: got %q, want the section path", got[0].Text)
	}
}

func TestClampIndexDoesNotWrap(t *testing.T) {
	cases := []struct{ current, delta, length, want int }{
		{0, -1, 5, 0},
		{4, 1, 5, 4},
		{0, 99, 5, 4},
		{4, -99, 5, 0},
		{1, 1, 5, 2},
	}
	for _, tc := range cases {
		if got := ClampIndex(tc.current, tc.delta, tc.length); got != tc.want {
			t.Errorf("ClampIndex(%d, %d, %d) = %d, want %d",
				tc.current, tc.delta, tc.length, got, tc.want)
		}
	}
}
